sys/chunked_ringbuffer/chunked_ringbuffer.c

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (C) 2021 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ *
+ * @author  Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include <string.h>
+#include "atomic_utils.h"
+#include "chunked_ringbuffer.h"
+#include "irq.h"
+
+static int _get_free_chunk(chunk_ringbuf_t *rb)
+{
+    int idx = rb->chunk_cur;
+    for (int i = 0; i < CHUNK_NUM_MAX; ++i) {
+        uintptr_t _ptr = atomic_load_uintptr((uintptr_t *)&rb->chunk_start[idx]);
+        if (_ptr == 0) {
+            return idx;
+        }
+
+        if (++idx == CHUNK_NUM_MAX) {
+            idx = 0;
+        }
+    }
+
+    return -1;
+}
+
+static int _get_complete_chunk(chunk_ringbuf_t *rb)
+{
+    int idx = rb->chunk_cur;
+    for (int i = 0; i < CHUNK_NUM_MAX; ++i) {
+        uintptr_t _ptr = atomic_load_uintptr((uintptr_t *)&rb->chunk_start[idx]);
+        if (_ptr) {
+            return idx;
+        }
+
```

Comment:
I *think* that _get_complete_chunk could do without this loop (if chunk_cur is advanced precisely once when protect is moved), but hey it works, and it might be over-optimization.